### PR TITLE
support raft learner in raftstore. 

### PR DESCRIPTION
**Pull request by hicqu**
This PR remove `unimplemented` panic for raft learn in raftstore by implement it.
Now we can do:
1. PeerStorage can know it's learner or not when it start up;
2. Report learners and pending-learners to pd;
3. Send AddLearner/PromoteLearner ConfChange to raftstore.

### CODE REVIEW

**Comment by siddontang** [comment]
please add a description

**Comment by hicqu** [comment]
@BusyJay @overvenus @siddontang PTAL thanks.

File: src/pd/mod.rs [hunk 1]
```diff
@@ -46,9 +47,11 @@ pub struct RegionStat {
 }
 
 impl RegionStat {
+    #[allow(too_many_arguments)]
```
Comment (BusyJay) [inline src/pd/mod.rs]: Any difference between using `new` and initialize it directly?

File: src/raft/raft.rs [hunk 2]
```diff
@@ -1838,6 +1838,10 @@ impl<T: Storage> Raft<T> {
         self.add_voter_or_learner(id, true);
     }
 
+    pub fn promote_learner(&mut self, id: u64) {
```
Comment (BusyJay) [inline src/raft/raft.rs]: What does it mean? Should it return error if the `id` is not learner?

File: src/raft/raw_node.rs [hunk 3]
```diff
@@ -293,6 +293,7 @@ impl<T: Storage> RawNode<T> {
             ConfChangeType::AddNode => self.raft.add_node(nid),
             ConfChangeType::AddLearnerNode => self.raft.add_learner(nid),
             ConfChangeType::RemoveNode => self.raft.remove_node(nid),
+            ConfChangeType::PromoteLearnerNode => self.raft.promote_learner(nid),
```
Comment (BusyJay) [inline src/raft/raw_node.rs]: I suggest to just reuse `AddNode` directly.

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -191,6 +191,7 @@ pub struct Peer {
     raft_engine: Arc<DB>,
     cfg: Rc<Config>,
     peer_cache: RefCell<FlatMap<u64, metapb::Peer>>,
+    learner_cache: RefCell<FlatMap<u64, metapb::Peer>>,
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: This is redundant. Use `peer_cache` instead.

File: src/raftstore/store/peer.rs [hunk 5]
```diff
@@ -1204,7 +1220,16 @@ impl Peer {
                     return Ok(());
                 }
             }
-            ConfChangeType::AddLearnerNode => unimplemented!(),
+            ConfChangeType::AddLearnerNode => {
+                let mut progress = Progress::default();
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: This is unnecessary.

File: src/raftstore/store/peer.rs [hunk 6]
```diff
@@ -1204,7 +1220,16 @@ impl Peer {
...
+                progress.is_learner = true;
+                status.learner_progress.insert(peer.get_id(), progress);
+                return Ok(());
+            }
+            ConfChangeType::PromoteLearnerNode => {
+                status.learner_progress.remove(&peer.get_id());
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why?

File: src/raftstore/store/peer.rs [hunk 7]
```diff
@@ -199,6 +200,7 @@ pub struct Peer {
     pending_reads: ReadIndexQueue,
     // Record the last instant of each peer's heartbeat response.
     pub peer_heartbeats: FlatMap<u64, Instant>,
+    pub learner_heartbeats: FlatMap<u64, Instant>,
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: Seem we can use peer_heartbeats only.

File: src/raftstore/store/peer.rs (hunk 4 above)
Comment (hicqu) [reply src/raftstore/store/peer.rs]: So should we rename `peer_cache` to `node_cache`? I think in our context, `peer` means it can vote.

Comment (BusyJay) [reply src/raftstore/store/peer.rs]: No, peer means `metapb::Peer`, which means a replica. Follower means it can vote.

File: src/raftstore/store/peer.rs (hunk 6 above)
Comment (hicqu) [reply src/raftstore/store/peer.rs]: We need `learner_progress` because it's used when we report `pending_learners` to pd.

File: src/raftstore/store/peer.rs (hunk 7 above)
Comment (hicqu) [reply src/raftstore/store/peer.rs]: In my opinion, `peer` means it can vote so that `learner` is not `peer`. So considering function [check_peers](https://github.com/hicqu/tikv/blob/3ebc84ee7b761bce7a27c782847c030e627f774c/src/raftstore/store/peer.rs#L548), I think seperate it is better.

**Comment by hicqu** [comment]
@BusyJay @siddontang PTAL thanks.

File: src/raftstore/store/peer.rs (hunk 7 above)
Comment (hicqu) [reply src/raftstore/store/peer.rs]: Maybe rename `peer_heartbeats` to `follower_heartbeats` is better?

File: tests/raftstore_cases/test_conf_change.rs [hunk 8]
```diff
@@ -637,3 +675,17 @@ fn test_conf_change_remove_leader() {
...
+
+#[test]
+fn test_node_learner_conf_change() {
+    let count = 5;
+    let mut cluster = new_node_cluster(0, count);
+    test_learner_conf_change(&mut cluster);
+}
+
+#[test]
+fn test_server_learner_conf_change() {
```
Comment (BusyJay) [inline tests/raftstore_cases/test_conf_change.rs]: Just node case is OK.

File: tests/raftstore_cases/test_conf_change.rs [hunk 9]
```diff
@@ -591,6 +591,44 @@ fn test_conf_change_safe<T: Simulator>(cluster: &mut Cluster<T>) {
...
+    // Add learner (4, 10) to region 1.
+    let engine_4 = cluster.get_engine(4);
+    pd_client.must_add_learner(r1, new_peer(4, 10));
+    cluster.must_put(b"k2", b"v2");
+    must_get_equal(&engine_4, b"k1", b"v1");
+    must_get_equal(&engine_4, b"k2", b"v2");
+
+    // Add peer on store which already has learner.
+    pd_client.add_peer(r1, new_peer(4, 10));
+    pd_client.must_none_peer(r1, new_peer(4, 10));
```
Comment (BusyJay) [inline tests/raftstore_cases/test_conf_change.rs]: How to ensure the command has been proposed and applied?

**Comment by hicqu** [comment]
@overvenus @BusyJay PTAL thanks.

File: src/raftstore/store/peer.rs [hunk 10]
```diff
@@ -1209,7 +1220,15 @@ impl Peer {
                     return Ok(());
                 }
             }
-            ConfChangeType::AddLearnerNode => unimplemented!(),
+            ConfChangeType::AddLearnerNode => {
+                return Ok(());
+            }
+            ConfChangeType::PromoteLearnerNode => {
+                if let Some(mut prog) = status.learner_progress.remove(&peer.get_id()) {
```
[the remark this answers is not in the document]
Comment (hicqu) [reply src/raftstore/store/peer.rs]: I think `remove` is more clear because we don't need consider `Entry::Vacant`.

**Comment by hicqu** [comment]
@overvenus PTAL again thanks.

File: src/raftstore/store/util.rs [hunk 11]
```diff
@@ -31,13 +31,21 @@ use super::engine::{IterOption, Iterable};
...
 
-    None
+pub fn find_learner(region: &metapb::Region, store_id: u64) -> Option<&metapb::Peer> {
+    region
+        .get_learners()
+        .iter()
+        .find(|p| p.get_store_id() == store_id)
+}
+
+pub fn find_peer_or_learner(region: &metapb::Region, store_id: u64) -> Option<&metapb::Peer> {
```
Comment (BusyJay) [inline src/raftstore/store/util.rs]: I think `find_peer` is OK. `peer` means replica, it's not the opposite of learner. The opposite of learner should be voter. Both learners and voters are peers.

File: src/raft/raw_node.rs [hunk 12]
```diff
@@ -290,7 +290,7 @@ impl<T: Storage> RawNode<T> {
         }
         let nid = cc.get_node_id();
         match cc.get_change_type() {
-            ConfChangeType::AddNode => self.raft.add_node(nid),
+            ConfChangeType::AddNode | ConfChangeType::PromoteLearnerNode => self.raft.add_node(nid),
```
[the remark this answers is not in the document]
Comment (hicqu) [reply src/raft/raw_node.rs]: I think change raft is also OK after `raft-rs` is used. That's very easy. I think `PromoteLearner` is more clear than `AddNode`. @Connor1996 what do you think?

[the remark this answers is not in the document]
Comment (BusyJay) [reply src/raft/raw_node.rs]: Since raft-rs doesn't handle them differently, distinguish them is meaningless and introduce complexity.

[the remark this answers is not in the document]
Comment (hicqu) [reply src/raft/raw_node.rs]: Got it. @Connor1996 if you agree too, please update https://github.com/pingcap/kvproto/pull/226.

**Comment by hicqu** [comment]
@BusyJay @overvenus PTAL again thanks.

File: src/pd/pd.rs [hunk 13]
```diff
@@ -54,6 +54,7 @@ pub enum Task {
         peer: metapb::Peer,
         down_peers: Vec<pdpb::PeerStats>,
```
Comment (BusyJay) [inline src/pd/pd.rs]: Does `down_learners` need to be considered?

Comment (hicqu) [reply src/pd/pd.rs]: fixed.

File: src/raftstore/store/peer.rs [hunk 14]
```diff
@@ -545,27 +543,38 @@ impl Peer {
...
+        let learners_count = self.region().get_learners().len();
+        if self.peer_heartbeats.len() == voters_count + learners_count {
             return;
         }
-
         // Insert heartbeats in case that some peers never response heartbeats.
-        for peer in self.region().get_peers().to_owned() {
+        for peer in self.region()
+            .get_peers()
+            .to_owned()
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: Looks like we do not need to `to_owned`.

File: src/raftstore/store/peer.rs [hunk 15]
```diff
@@ -545,27 +543,38 @@ impl Peer {
...
                 .entry(peer.get_id())
                 .or_insert_with(Instant::now);
         }
     }
 
-    pub fn collect_down_peers(&self, max_duration: Duration) -> Vec<PeerStats> {
+    pub fn collect_down_peers(&self, max_duration: Duration, is_learner: bool) -> Vec<PeerStats> {
         let mut down_peers = Vec::new();
-        for p in self.region().get_peers() {
+        let peers = if !is_learner {
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: How about `if is_learner { learner } else { peer }`?

File: src/raftstore/store/worker/apply.rs [hunk 16]
```diff
@@ -771,13 +771,14 @@ impl ApplyDelegate {
                 }
 
                 // TODO: Do we allow adding peer in same node?
-
                 region.mut_peers().push(peer.clone());
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: What if the learner does not exist?

**Comment by hicqu** [comment]
I have made a refactor to unify `peers` and `learners` in `Region`. PTAL again, thanks. @overvenus @BusyJay .

File: tests/raftstore_cases/test_conf_change.rs [hunk 17]
```diff
@@ -591,6 +591,40 @@ fn test_conf_change_safe<T: Simulator>(cluster: &mut Cluster<T>) {
...
+    must_get_equal(&engine_4, b"k2", b"v2");
+
+    // Remove learner (4, 10) from region 1.
+    pd_client.must_remove_peer(r1, new_peer(4, 10));
+    pd_client.must_add_peer(r1, new_learner_peer(4, 10));
+    must_get_equal(&engine_4, b"k2", b"v2");
+
+    // Promote learner (4 ,10) to voter.
+    pd_client.must_add_peer(r1, new_peer(4, 10));
+    cluster.stop_node(2);
```
Comment (overvenus) [inline tests/raftstore_cases/test_conf_change.rs]: Why stop 2?

File: src/raftstore/store/peer.rs [hunk 18]
```diff
@@ -1255,15 +1266,27 @@ impl Peer {
...
+                    // For promote learner to voter.
+                    progress.is_learner = false;
+                    status.progress.insert(peer.get_id(), progress);
+                } else {
+                    status.progress.insert(peer.get_id(), Progress::default());
+                }
             }
             ConfChangeType::RemoveNode => {
+                if peer.get_is_learner() {
+                    // If the node is a learner, we can return directly.
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: Do we need to clean the `learner_progress`?

File: src/raftstore/store/peer.rs [hunk 19]
```diff
@@ -1255,15 +1266,27 @@ impl Peer {
...
+                    return Ok(());
+                }
                 if status.progress.remove(&peer.get_id()).is_none() {
                     // It's always safe to remove a unexisting node.
                     return Ok(());
                 }
             }
-            ConfChangeType::AddLearnerNode => unimplemented!(),
+            ConfChangeType::AddLearnerNode => {
+                return Ok(());
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: Does it mean it should not be reachable?

File: src/raftstore/store/util.rs [hunk 20]
```diff
@@ -367,6 +379,18 @@ pub fn is_sibling_regions(lhs: &metapb::Region, rhs: &metapb::Region) -> bool {
     false
 }
 
+pub fn conf_state_from_region(region: &metapb::Region) -> ConfState {
```
Comment (siddontang) [inline src/raftstore/store/util.rs]: add test for it.

File: src/raftstore/store/peer.rs (hunk 18 above)
Comment (hicqu) [reply src/raftstore/store/peer.rs]: The `learner_progress` here is seperated from `RawNode` because `RawNode::status` has already copyed the field. So we don't do that here.

File: src/raftstore/store/peer.rs (hunk 19 above)
Comment (hicqu) [reply src/raftstore/store/peer.rs]: No, it's reachable. `return OK` means `AddLearner` won't break the safe rule of conf change, described at line 1222.

File: src/raftstore/store/util.rs (hunk 20 above)
Comment (hicqu) [reply src/raftstore/store/util.rs]: fixed.

File: src/raftstore/store/worker/apply.rs [hunk 21]
```diff
@@ -960,26 +958,31 @@ impl ApplyDelegate {
...
+                            "can't add duplicated peer {:?} to region {:?}",
+                            peer,
+                            self.region
+                        ));
+                    } else {
+                        p.set_is_learner(false);
+                    }
+                }
+                if !exists {
+                    // TODO: Do we allow adding peer in same node?
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: For now, we do not allow same peer in same node, can we remove it?

File: src/raftstore/store/worker/apply.rs [hunk 22]
```diff
@@ -990,7 +993,24 @@ impl ApplyDelegate {
                     .with_label_values(&["remove_peer", "all"])
                     .inc();
 
-                if !exists {
+                if let Some(p) = util::remove_peer(&mut region, store_id) {
+                    if &p != peer {
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: It looks like `is_learner` sensitive, could you add some comments?

File: tests/raftstore_cases/test_conf_change.rs [hunk 23]
```diff
@@ -591,6 +591,39 @@ fn test_conf_change_safe<T: Simulator>(cluster: &mut Cluster<T>) {
...
+
+    // Add learner (4, 10) to region 1.
+    let engine_4 = cluster.get_engine(4);
+    pd_client.must_add_peer(r1, new_learner_peer(4, 10));
+    cluster.must_put(b"k2", b"v2");
+    must_get_equal(&engine_4, b"k1", b"v1");
+    must_get_equal(&engine_4, b"k2", b"v2");
+
+    // Remove learner (4, 10) from region 1.
+    pd_client.must_remove_peer(r1, new_peer(4, 10));
```
Comment (overvenus) [inline tests/raftstore_cases/test_conf_change.rs]: It should be `new_learner_peer(4, 10)`.

File: tests/raftstore/pd.rs [hunk 24]
```diff
@@ -766,12 +776,12 @@ impl TestPdClient {
...
     pub fn must_add_peer(&self, region_id: u64, peer: metapb::Peer) {
         self.add_peer(region_id, peer.clone());
-        self.must_have_peer(region_id, peer);
+        self.must_have_peer(region_id, peer.clone());
     }
 
     pub fn must_remove_peer(&self, region_id: u64, peer: metapb::Peer) {
         self.remove_peer(region_id, peer.clone());
-        self.must_none_peer(region_id, peer);
+        self.must_none_peer(region_id, peer.clone());
```
Comment (overvenus) [inline tests/raftstore/pd.rs]: We don't need to call clone.

File: tests/raftstore/pd.rs [hunk 25]
```diff
@@ -694,37 +706,35 @@ impl TestPdClient {
...
                 Some(region) => region,
                 None => continue,
             };
 
             if let Some(p) = find_peer(&region, peer.get_store_id()) {
-                if p.get_id() == peer.get_id() {
+                if p == &peer {
                     return;
                 }
+                continue;
```
Comment (overvenus) [inline tests/raftstore/pd.rs]: It looks like redundant.

File: tests/raftstore/pd.rs [hunk 26]
```diff
@@ -588,6 +599,7 @@ fn must_same_peers(left: &metapb::Region, right: &metapb::Region) {
     for peer in left.get_peers() {
         let p = find_peer(right, peer.get_store_id()).unwrap();
         assert_eq!(p.get_id(), peer.get_id());
+        assert_eq!(p.get_is_learner(), peer.get_is_learner());
```
Comment (overvenus) [inline tests/raftstore/pd.rs]: Just `assert!(p, peer)`

File: src/raftstore/store/util.rs [hunk 27]
```diff
@@ -32,13 +32,17 @@ use super::engine::{IterOption, Iterable};
...
-        }
-    }
+    region
+        .get_peers()
+        .iter()
+        .find(|&p| p.get_store_id() == store_id)
+}
 
-    None
+pub fn find_mut_peer(region: &mut metapb::Region, store_id: u64) -> Option<&mut metapb::Peer> {
```
Comment (overvenus) [inline src/raftstore/store/util.rs]: Could you rename it to `find_peer_mut()`?

File: src/pd/pd.rs [hunk 28]
```diff
@@ -558,15 +558,15 @@ impl<T: PdClient> Runnable<Task> for Runner<T> {
...
-                    RegionStat::new(
-                        down_peers,
-                        pending_peers,
-                        written_bytes_delta,
-                        written_keys_delta,
-                        read_bytes_delta,
-                        read_keys_delta,
-                        approximate_size,
-                    ),
+                    RegionStat {
```
Comment (overvenus) [inline src/pd/pd.rs]: Unrelated changes.

Comment (hicqu) [reply src/pd/pd.rs]: I prefer keep the new code because add new arguments to `RegionStat::new` will cause `TooManyArguments` lint warning.

File: src/raftstore/store/worker/apply.rs (hunk 21 above)
Comment (hicqu) [reply src/raftstore/store/worker/apply.rs]: I agree. @BusyJay please also check it.

Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: Not yet. It may in the future.

File: Cargo.toml [hunk 29]
```diff
@@ -82,7 +82,8 @@ git = "https://github.com/pingcap/murmur3.git"
 git = "https://github.com/pingcap/rust-rocksdb.git"
 
 [dependencies.kvproto]
-git = "https://github.com/pingcap/kvproto.git"
```
Comment (Hoverbear) [inline Cargo.toml]: Please remember to swap the branch back before merging. :)

File: src/pd/pd.rs (hunk 28 above)
Comment (Hoverbear) [reply src/pd/pd.rs]: I suspect you can do `#[allow(TooManyArguments)]` on this block if that's the case.

File: src/raftstore/store/peer.rs [hunk 30]
```diff
@@ -572,7 +569,8 @@ impl Peer {
         }
 
         // Insert heartbeats in case that some peers never response heartbeats.
-        for peer in self.region().get_peers().to_owned() {
+        let region = self.raft_group.get_store().get_region();
```
Comment (Hoverbear) [inline src/raftstore/store/peer.rs]: Superfluous change but it's okay. :)

File: src/raftstore/store/util.rs [hunk 31]
```diff
@@ -32,13 +32,17 @@ use super::engine::{IterOption, Iterable};
 use super::peer_storage;
 
 pub fn find_peer(region: &metapb::Region, store_id: u64) -> Option<&metapb::Peer> {
-    for peer in region.get_peers() {
-        if peer.get_store_id() == store_id {
-            return Some(peer);
-        }
-    }
+    region
```
Comment (Hoverbear) [inline src/raftstore/store/util.rs]: Beautiful. 👍

File: src/raftstore/store/worker/apply.rs [hunk 32]
```diff
@@ -960,26 +958,31 @@ impl ApplyDelegate {
...
+                        );
+                        return Err(box_err!(
+                            "can't add duplicated peer {:?} to region {:?}",
+                            peer,
+                            self.region
+                        ));
+                    } else {
+                        p.set_is_learner(false);
+                    }
+                }
```
Comment (Hoverbear) [inline src/raftstore/store/worker/apply.rs]: I believe the pattern of

```rust
if let ... {
  // ...
} else {
  // ...
}
```

If preferred over

```rust
let mut flag = false;
if let ... {
  flag = true;
  // ...
}
if flag {
  // ...
}
```

Comment (hicqu) [reply src/raftstore/store/worker/apply.rs]: yes, but `if let` will borrow something so that we can't borrow it again in `else` branch. It's a lifetime issue. So I will comment it.

Comment (Hoverbear) [reply src/raftstore/store/worker/apply.rs]: Ah! This makes sense. Thanks for clarifying.

File: tests/raftstore_cases/test_conf_change.rs [hunk 33]
```diff
@@ -591,6 +591,47 @@ fn test_conf_change_safe<T: Simulator>(cluster: &mut Cluster<T>) {
...
+    must_get_equal(&engine_4, b"k1", b"v1");
+    must_get_equal(&engine_4, b"k2", b"v2");
+
+    // Remove learner (4, 10) from region 1.
+    pd_client.must_remove_peer(r1, new_learner_peer(4, 10));
+    pd_client.must_add_peer(r1, new_learner_peer(4, 10));
+    must_get_equal(&engine_4, b"k2", b"v2");
+
+    // Promote learner (4 ,10) to voter.
+    pd_client.must_add_peer(r1, new_peer(4, 10));
```
Comment (overvenus) [inline tests/raftstore_cases/test_conf_change.rs]: How do we ensure it becomes a voter?

Comment (hicqu) [reply tests/raftstore_cases/test_conf_change.rs]: `must_add_peer` will check it's learner or voter when receiving heartbeats.
